File: Snooper-Spring/src/main/java/org/springframework/http/client/SnooperInterceptor.java
package org.springframework.http.client;

import com.prateekj.snooper.AndroidSnooper;
import com.prateekj.snooper.networksnooper.model.HttpCall;

import org.springframework.http.HttpRequest;

import java.io.IOException;

public class SnooperInterceptor implements ClientHttpRequestInterceptor {

  @Override
  public ClientHttpResponse intercept(HttpRequest request, byte[] byteArray,
                                      ClientHttpRequestExecution execution) throws IOException {

    SpringHttpRequestTransformer transformer = new SpringHttpRequestTransformer();
    AndroidSnooper snooper = AndroidSnooper.getInstance();
    ClientHttpResponse streamResponse;
    try {
      streamResponse = execution.execute(request, byteArray);
    } catch (Exception e) {
      HttpCall call = transformer.transform(request, byteArray, e);
      snooper.record(call);
      throw e;
    }
    BufferingClientHttpResponseWrapper httpResponse = new BufferingClientHttpResponseWrapper(streamResponse);
    HttpCall call = transformer.transform(request, byteArray, httpResponse);
    snooper.record(call);
    return httpResponse;
  }
}

File: Snooper-Spring/src/main/java/org/springframework/http/client/SnooperInterceptor.kt
package org.springframework.http.client

import com.prateekj.snooper.AndroidSnooper
import org.springframework.http.HttpRequest
import java.io.IOException

class SnooperInterceptor : ClientHttpRequestInterceptor {

  @Throws(IOException::class)
  override fun intercept(
    request: HttpRequest, byteArray: ByteArray,
    execution: ClientHttpRequestExecution
  ): ClientHttpResponse {

    val transformer = SpringHttpRequestTransformer()
    val snooper = AndroidSnooper.instance
    val streamResponse: ClientHttpResponse
    try {
      streamResponse = execution.execute(request, byteArray)
    } catch (e: Exception) {
      val call = transformer.transform(request, byteArray, e)
      snooper.record(call)
      throw e
    }

    val httpResponse = BufferingClientHttpResponseWrapper(streamResponse)
    val call = transformer.transform(request, byteArray, httpResponse)
    snooper.record(call)
    return httpResponse
  }
}